### PR TITLE
Stop writing unparseable requirements to Pipfiles

### LINKED ISSUE #1056: pipenv: error: no such option: --upgrade
Getting an no such option error with every install command.

##### Describe your environment

1. MacOS High Sierra
1. Python version: 2.7
1. Pipenv version: 8.3.2

##### Expected result

Just install the package

##### Actual result

```
$ pipenv install --verbose
Pipfile.lock not found, creating…
Locking [dev-packages] dependencies…
Locking [packages] dependencies…
Usage: pipenv [options]

Traceback (most recent call last):
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/bin/pipenv", line 9, in <module>
    load_entry_point('pipenv==8.3.2', 'console_scripts', 'pipenv')()
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/vendor/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/vendor/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/vendor/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/vendor/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/vendor/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/cli.py", line 1866, in install
    do_init(dev=dev, allow_global=system, ignore_pipfile=ignore_pipfile, system=system, skip_lock=skip_lock, verbose=verbose, concurrent=concurrent, deploy=deploy, pre=pre)
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/cli.py", line 1330, in do_init
    do_lock(system=system, pre=pre)
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/cli.py", line 1121, in do_lock
    pre=pre
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/utils.py", line 426, in resolve_deps
    constraint = [c for c in pip.req.parse_requirements(t, session=pip._vendor.requests)][0]
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/patched/pip/req/req_file.py", line 93, in parse_requirements
    for req in req_iter:
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/patched/pip/req/req_file.py", line 140, in process_line
    opts, _ = parser.parse_args(shlex.split(options_str), defaults)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/optparse.py", line 1402, in parse_args
    self.error(str(err))
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/optparse.py", line 1584, in error
    self.exit(2, "%s: error: %s\n" % (self.get_prog_name(), msg))
  File "/Users/aleskocur-rubicoin/Library/Python/2.7/lib/python/site-packages/pipenv/patched/pip/req/req_file.py", line 284, in parser_exit
    raise RequirementsFileParseError(msg)
pip.exceptions.RequirementsFileParseError: pipenv: error: no such option: --upgrade
```

##### Steps to replicate

It started when I run by mistake `pipenv install [package] --upgrade`.

### PR DESCRIPTION
**Pull request by techalchemy**
- Add better exception handling and error messaging for parse errors
- Add test to verify this is now being handled correctly
- Fixes #1056

### CODE REVIEW

**Comment by nateprewitt** [comment]
We should just be able to do a single split on the first space, no? That's should give us a clear break in the package name.

I meant to add that in the last chunk of fixes I did but it never made it in.

**Comment by techalchemy** [comment]
Not sure, I'm not really testing for that... I didn't change any of that code, only added exception handling

**Comment by nateprewitt** [comment]
The exception is raised because of that cluttered line though, right? If we fix the call so the exception won't happen then we don't need to unecessarily warn.

**Comment by techalchemy** [comment]
The use case presented just happened to try to use `--upgrade` as a thing they were installing, but it would go similarly if you tried to `pipenv install abc{+/-\||<><{}}}}#@%$#@^` or some other invalid requirement (that isn't very hard, since the requirements parser throws an exception on anything that doens't conform to pep440)

**Comment by techalchemy** [comment]
No the exception is raised because the user passed in an invalid package name and the requirements parser can't parse it so it throws one

**Comment by techalchemy** [comment]
Current Pipenv:

```console
 ◰³ tempenv-6eb9235674bb2  /t/new  pipenv install invali@!+packgg
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
Creating a Pipfile for this project…
Installing invali@!+packgg…
Traceback (most recent call last):
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3015, in __init__
    super(Requirement, self).__init__(requirement_string)
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pkg_resources/_vendor/packaging/requirements.py", line 101, in __init__
    raise InvalidRequirement("Invalid URL given")
pkg_resources.extern.packaging.requirements.InvalidRequirement: Invalid URL given

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/bin/pipenv", line 11, in <module>
    sys.exit(cli())
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pipenv/vendor/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pipenv/cli.py", line 1876, in install
    c = pip_install(package_name, ignore_hashes=True, allow_global=system, no_deps=False, verbose=verbose, pre=pre)
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pipenv/cli.py", line 1355, in pip_install
    if [x for x in requirements.parse(package_name.split('--hash')[0].split('--trusted-host')[0])][0].vcs:
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pipenv/cli.py", line 1355, in <listcomp>
    if [x for x in requirements.parse(package_name.split('--hash')[0].split('--trusted-host')[0])][0].vcs:
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pipenv/vendor/requirements/parser.py", line 50, in parse
    yield Requirement.parse(line)
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pipenv/vendor/requirements/requirement.py", line 220, in parse
    return cls.parse_line(line)
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pipenv/vendor/requirements/requirement.py", line 198, in parse_line
    pkg_req = Req.parse(line)
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3061, in parse
    req, = parse_requirements(s)
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3008, in parse_requirements
    yield Requirement(line)
  File "/home/hawk/.virtualenvs/tempenv-6eb9235674bb2/lib/python3.6/site-packages/pkg_resources/__init__.py", line 3017, in __init__
    raise RequirementParseError(str(e))
pkg_resources.RequirementParseError: Invalid URL given
```

With this PR:

```console
 ◰³ tempenv-4b7698643737  /t/new  pipenv install invali@!+packgg                                                                                                                  Tue 14 Nov 2017 09:42:35 PM EST
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
Creating a Pipfile for this project…
Installing invali@!+packgg…
WARNING: Invalid URL given
ABORTING INSTALL... You will have to reinstall any packages that failed to install.
⠋You may have to manually run pipenv lock when you are finished.
```

File: tests/test_pipenv.py [hunk 1]
```diff
@@ -174,6 +174,14 @@ def test_man(self):
             c = p.pipenv('--man')
             assert c.return_code == 0 or c.err
 
+    @pytest.mark.cli
+    @pytest.mark.install
+    def test_install_parse_error(self):
+        with PipenvInstance() as p:
+            c = p.pipenv('install tablib --upgrade')
```
Comment (vphilippon) [inline tests/test_pipenv.py]: I would suggest to put a comment here to explain that the invalid input is `--upgrade` here. The reason of the parse error is not clear at first.
(I'm sure our future selves could forget about that one :) )

Comment (nateprewitt) [reply tests/test_pipenv.py]: Yep, a comment is a good call. Then I think we're set to merge here. We'll take care of the option issue in another PR.

**Comment by techalchemy** [comment]
I could just be less confusing and use clearly unparseable package inputs that won’t be confused for options :)

**Comment by techalchemy** [comment]
Let's see what the CI gods have to say

**Comment by nateprewitt** [comment]
The CI gods have chosen to smile favorably upon us this day. 🙏

**Pull request merged**